Zend/zend_builtin_functions.c

```diff
@@ -491,12 +491,16 @@ ZEND_FUNCTION(func_num_args)
 {
 	zend_execute_data *ex = EX(prev_execute_data);
 
-	if (!(ZEND_CALL_INFO(ex) & ZEND_CALL_CODE)) {
-		RETURN_LONG(ZEND_CALL_NUM_ARGS(ex));
-	} else {
+	if (ZEND_CALL_INFO(ex) & ZEND_CALL_CODE) {
 		zend_error(E_WARNING, "func_num_args():  Called from the global scope - no function context");
 		RETURN_LONG(-1);
 	}
+
+	if (zend_forbid_dynamic_call("func_num_args()") == FAILURE) {
```

Comment:
@laruence Ahhh, sorry, I didn't catch what you mean. The reason why I'm passing in the function name is to throw an error like "Cannot call parse_str() with a single argument dynamically" for parse_str/mb_parse_str, as these are only forbidden if the second argument is not used. If I remove this special case we can indeed determine the function name automatically.